### ISSUE TITLE
question about prior model

### ISSUE THREAD

**Issue by meadow163** [acids-ircam/RAVE#68]
Hi, caillonantoine

In your paper, you said 'To illustrate this, we train a WaveNet-inspired model at generating latent signals in an autoregressive
fashion for performing audio synthesis with RAVE', I don't understand your prior model, in your code:

```python
x = self.encode(batch) # here got the Z code
x = self.quantized_normal.encode(self.diagonal_shift(x)) # here what did you mean? why do you need it? 
pred = self.forward(x)

x = torch.argmax(self.split_classes(x[..., 1:]), -1)
pred = self.split_classes(pred[..., :-1]) # if here predicts next z code, why did you use cross_entropy loss function?

loss = nn.functional.cross_entropy(
    pred.reshape(-1, self.quantized_normal.resolution),
    x.reshape(-1),
)
```
could you give me some reference paper or code which you refer?
Best ,

**Comment by meadow163** [comment]
I train the prior model like following code:

`   losses = []
    global_step = epoch * len(train_loader)
    # avg_batch = True
    for i, batch in enumerate(train_loader):

        # x = data.cuda()
        optimizer.zero_grad()
        with torch.no_grad():
            x = batch.unsqueeze(1).cuda()
            for encoder in vae_model.generator:
                x = encoder(x)
            h = vae_model.mid_layer(x)
            mu = vae_model.fc1(h.transpose(1, 2))
            logvar = vae_model.fc2(h.transpose(1, 2))
            z = vae_model.reparameterize(mu, logvar)

        x = prior_model.quantized_normal.encode(prior_model.diagonal_shift(z.permute(0,2,1)))
        pred = prior_model.forward(x)

        x = torch.argmax(prior_model.split_classes(x[..., 1:]), -1)
        pred = prior_model.split_classes(pred[..., :-1])

        loss = nn.functional.cross_entropy(
            pred.reshape(-1, prior_model.quantized_normal.resolution),
            x.reshape(-1),
        )

        loss.backward()
        # total_loss.backward()
        clip_grad_norm_(prior_model.parameters(),max_norm=3,norm_type=2)
        optimizer.step()

        mmse = loss.data.cpu().item()
        losses.append(mmse)`

the training curve is like this:
![image](https://user-images.githubusercontent.com/30004294/169485187-93ada21f-ccb0-4749-b914-d8f035f6756e.png)
I think it  is convergent, But when I try to generate as following code:
` data  = data[0:262144]
        data = torch.FloatTensor(data)
        # data = data.unsqueeze(0)

        with torch.no_grad():

            x = data.unsqueeze(0).unsqueeze(0).cuda()
            for encoder in vae_model.generator:
                x = encoder(x)

            h = vae_model.mid_layer(x)
            mu = vae_model.fc1(h.transpose(1, 2))
            logvar = vae_model.fc2(h.transpose(1, 2))
            z = vae_model.reparameterize(mu, logvar)  #
            z_reco = z

            # prior
            # here differs with training:prior_model.quantized_normal.encode(prior_model.diagonal_shift(z.permute(0,2,1)))
            z = prior_model.quantized_normal.encode(z.permute(0,2,1))
            z = prior_model.generate(z)
            z = prior_model.diagonal_shift.inverse(prior_model.quantized_normal.decode(z))

            x = vae_model.fc3(z.permute(0,2,1)).transpose(1, 2)
            for decoder in vae_model.decoder:
                x = decoder(x)

            gen_wav = x
            reco = vae_model.fc3(z_reco).transpose(1, 2)
            for decoder in vae_model.decoder:
                reco = decoder(reco)

            reco = reco.cpu().squeeze().numpy()
            gen_wav = gen_wav.cpu().squeeze().numpy()
            data = data.squeeze().cpu().numpy()

            sf.write(os.path.join(args['outdir'], name.replace('.wav', '') + f'epoch{epoch}_reco.wav'), reco, sr)
            sf.write(os.path.join(args['outdir'],name.replace('.wav','')+f'epoch{epoch}_gen.wav'),gen_wav,sr)
            sf.write(os.path.join(args['outdir'],name),data,sr)`
reconstruction wave:
![image](https://user-images.githubusercontent.com/30004294/169486306-45a215a0-81af-42db-8259-fe81ce5b84f7.png)
generation wave:
![image](https://user-images.githubusercontent.com/30004294/169486404-8f0d58b6-f22f-447c-8906-8f3c5a1a2454.png)

the generation wave like noisy, I think there are some issues with it.

my train environment has no pytorch_lightning, so I change the prior model like:

`import torch
import torch.nn as nn
# import pytorch_lightning as pl
from tqdm import tqdm

from models.prior.residual_block import ResidualBlock
from models.prior.core import DiagonalShift, QuantizedNormal

# import cached_conv as cc


def get_padding(kernel_size, stride=1, dilation=1, mode="centered"):

    if kernel_size == 1: return (0, 0)
    p = (kernel_size - 1) * dilation + 1
    half_p = p // 2
    if mode == "centered":
        p_right = half_p
        p_left = half_p
    elif mode == "causal":
        p_right = 0
        p_left = 2 * half_p
    elif mode == "anticausal":
        p_right = 2 * half_p
        p_left = 0
    else:
        raise Exception(f"Padding mode {mode} is not valid")
    # return (p_left, p_right)
    # return p_left+p_right
    # return p
    return half_p
# wavenet
class Model(nn.Module):

    def __init__(self, resolution=32, res_size=512,skp_size=256, kernel_size=3,
                 cycle_size=4,n_layers=10,data_size=64):
        super().__init__()
        # self.save_hyperparameters()

        self.diagonal_shift = DiagonalShift()
        self.quantized_normal = QuantizedNormal(resolution)

        # self.synth = torch.jit.load(pretrained_vae)
        self.synth = None

        # self.sr = self.synth.sampling_rate.item()
        # data_size = self.synth.cropped_latent_size

        self.pre_net = nn.Sequential(
            nn.Conv1d(
                resolution * data_size,
                res_size,
                kernel_size,
                padding=get_padding(kernel_size, mode="causal"),
                groups=data_size,
            ),
            nn.LeakyReLU(.2),
        )

        self.residuals = nn.ModuleList([
            ResidualBlock(
                res_size,
                skp_size,
                kernel_size,
                2**(i % cycle_size),
            ) for i in range(n_layers)
        ])

        self.post_net = nn.Sequential(
            nn.Conv1d(skp_size, skp_size, 1),
            nn.LeakyReLU(.2),
            nn.Conv1d(
                skp_size,
                resolution * data_size,
                1,
                groups=data_size,
            ),
        )

        self.data_size = data_size

        self.val_idx = 0

    def configure_optimizers(self):
        p = []
        p.extend(list(self.pre_net.parameters()))
        p.extend(list(self.residuals.parameters()))
        p.extend(list(self.post_net.parameters()))
        return torch.optim.Adam(p, lr=1e-4)

    @torch.no_grad()
    def encode(self, x):
        self.synth.eval()
        return self.synth.encode(x)

    @torch.no_grad()
    def decode(self, z):
        self.synth.eval()
        return self.synth.decode(z)

    def forward(self, x):
        res = self.pre_net(x)
        skp = torch.tensor(0.).to(x)
        for layer in self.residuals:
            res, skp = layer(res, skp)
        x = self.post_net(skp)
        return x

    @torch.no_grad()
    def generate(self, x, argmax: bool = False):
        for i in tqdm(range(x.shape[-1] - 1)):
            USE_BUFFER_CONV = True
            # if cc.USE_BUFFER_CONV:
            if USE_BUFFER_CONV:
                start = i
            else:
                start = None

            pred = self.forward(x[..., start:i + 1])

            # if not cc.USE_BUFFER_CONV:
            if not USE_BUFFER_CONV:
                pred = pred[..., -1:]

            pred = self.post_process_prediction(pred, argmax=argmax)

            x[..., i + 1:i + 2] = pred
        return x

    def split_classes(self, x):
        # B x D*C x T
        x = x.permute(0, 2, 1)
        x = x.reshape(x.shape[0], x.shape[1], self.data_size, -1)
        x = x.permute(0, 2, 1, 3)  # B x D x T x C
        return x

    def post_process_prediction(self, x, argmax: bool = False):
        x = self.split_classes(x)
        shape = x.shape[:-1]
        x = x.reshape(-1, x.shape[-1])

        if argmax:
            x = torch.argmax(x, -1)
        else:
            x = torch.softmax(x - torch.logsumexp(x, -1, keepdim=True), -1)
            x = torch.multinomial(x, 1, True).squeeze(-1)

        x = x.reshape(shape[0], shape[1], shape[2])
        x = self.quantized_normal.to_stack_one_hot(x)
        return x
`
Is there any issue with it? I don't know why I get the wrong result.
any reply will be appreciated, thank in advance!

**Comment by caillonantoine** [comment]
Hi, I suggest that you install the missing dependencies in order to use the original source code ! Concerning your first question you can read [this article](https://arxiv.org/abs/1802.08435) !
